harry_potter/spec/harry_potter_spec.rb

```diff
@@ -0,0 +1,156 @@
+module Shop
+  BASE_PRICE = 8
+
+  DISCOUNTS = {
+    1 => 0,
+    2 => 5,
+    3 => 10,
+    4 => 20,
+    5 => 25,
+  }
+
+  DISCOUNT_TIERS = DISCOUNTS.size
+
+  class Book
+    attr_accessor :price
+
+    def initialize(price)
+      @price = price
+    end
+
+    def to_s
+      "<Book price='#{@price}'>"
+    end
+  end
+
+  class BookList
+    attr_accessor :books
+    attr_reader   :sets
+
+    def initialize( books = [] )
+      @books = books
+    end
+
+    def method_missing(method, *args, &block)
+      if @books.respond_to? method
+        return @books.send(method, *args, &block)
+      end
+
+      super
+    end
+
+    def price
+      base_price = Shop::BASE_PRICE * @books.size
+
+      base_price - (base_price * discount)
+    end
+
+    def tier_prices
+      sets.map { |_, v| v.map(&:price).inject(:+) }
+    end
+
+    def best_price
+      tier_prices.min
+    end
+
+    def sets
+      @sets ||= find_all_sets(@books)
+    end
+
+    def set?
+      @books.uniq.size == @books.size
+    end
+
+    def discount
+      if set? then Shop::DISCOUNTS[@books.size] * 0.01 else 0 end
+    end
+
+    def to_s
+      "<BookList books='#{@books}'>"
+    end
+
+    private
+
+    def find_all_sets( books )
+      tier_sets = Hash.new(BookList.new())
```

Comment:
Did you mean to use the same instance of a BookList for unknown keys? I'm not sure its a problem here, but is there potential for a bug to creep in here?